src/inc/til/rle.h

```diff
@@ -0,0 +1,1064 @@
+            // * 0 <= start_index < _total_length
+            // * 0 < end_index <= _total_length
+            // * start_index < end_index
+            //
+            // --> It's safe to subtract 1 from end_index
+
+            rle_scanner scanner(_runs.begin(), _runs.end());
+            auto [begin_run, start_run_pos] = scanner.scan(start_index);
+            auto [end_run, end_run_pos] = scanner.scan(end_index - 1);
+
+            container slice{ begin_run, end_run + 1 };
+            slice.back().length = end_run_pos + 1;
+            slice.front().length -= start_run_pos;
+
+            return { std::move(slice), static_cast<size_type>(end_index - start_index) };
+        }
+
+        // Set the range [start_index, end_index) to the given value.
+        void replace(size_type start_index, size_type end_index, const value_type& value)
+        {
+            _check_indices(start_index, end_index);
+
+            const rle_type replacement{ value, static_cast<size_type>(end_index - start_index) };
+            _replace_unchecked(start_index, end_index, { &replacement, 1 });
+        }
+
+        // Replace the range [start_index, end_index) with the given run.
+        // NOTE: This can change the size/length of the vector.
+        void replace(size_type start_index, size_type end_index, const rle_type& replacement)
+        {
+            replace(start_index, end_index, { &replacement, 1 });
+        }
+
+        void replace(size_type start_index, size_type end_index, const gsl::span<const rle_type> replacements)
+        {
+            _check_indices(start_index, end_index);
+            _replace_unchecked(start_index, end_index, replacements);
+        }
+
+        // Replaces every value seen in the run with a new one
+        // Does not change the length or position of the values.
+        void replace_values(const value_type& old_value, const value_type& new_value)
+        {
+            for (auto& run : _runs)
+            {
+                if (run.value == old_value)
+                {
+                    run.value = new_value;
+                }
+            }
+
+            _compact();
+        }
+
+        // Adjust the size of the vector.
+        // If the size is being increased, the last run is extended to fill up the new vector size.
+        // If the size is being decreased, the trailing runs are cut off to fit.
+        void resize_trailing_extent(const size_type new_size)
+        {
+            if (new_size == 0)
+            {
+                _runs.clear();
+            }
+            else if (new_size < _total_length)
+            {
+                rle_scanner scanner(_runs.begin(), _runs.end());
+                auto [run, pos] = scanner.scan(new_size - 1);
+
+                run->length = ++pos;
+
+                _runs.erase(++run, _runs.cend());
+            }
+            else if (new_size > _total_length)
+            {
+                Expects(!_runs.empty());

```

Comment:
What is `Expects`? Is that some sort of assert? How does it fail?

Reply:
Expects is part of GSL, which I believe is a better fit for such generic contains than WIL.
It basically works the same as the other WIL assertion functions and aborts the program on failure (even in release builds).

Reply:
Works for me, thanks. Haven't seen it before.